### PR TITLE
standardize keybindings framework

### PR DESCRIPTION
**Pull request by kne42**
# Description

Add machinery for the adding, validation, and dispatch of keybindings on the viewer and its layers.

Supersedes #251.

### key combinations
Keys are represented in the form `[modifier-]key`, e.g. `a`,
`Control-c`, or `Control-Alt-Delete`.
Valid modifiers are Control, Alt, Shift, and Meta.

Letters will always be read as upper-case.
Due to the native implementation of the key system, Shift pressed in certain
key combinations may yield inconsistent or unexpected results.
Therefore, it is not recommended to use Shift with non-letter keys.
On OSX, Control is swapped with Meta such that pressing Command reads as
Control.

Special keys include Shift, Control, Alt, Meta, Up, Down, Left, Right,
PageUp, PageDown, Insert, Delete, Home, End, Escape, Backspace, F1,
F2, F3, F4, F5, F6, F7, F8, F9, F10, F11, F12, Space, Enter, and Tab

### function format
Functions take in only one argument: the parent that the function was bound to. This is the viewer or layer.

By default, all functions are assumed to work on key presses only, but can be denoted to work on release too by separating the function into two statements with the `yield` keyword:
```python
def hello_world(viewer):
    # on key press
    viewer.status = 'hello world!'

    yield

    # on key release
    viewer.status = 'goodbye world :('
```

### bind a key
```python
viewer.bind_key('n', load_next_image)

@viewer.bind_key('p')
def load_prev_image(viewer): ...
```

### unbind a key
```python
viewer.bind_key('p', None)
```

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# References

see #210, #251

# How has this been tested?

- [x] the test suite for my feature passes
- [x] all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by sofroniewn — APPROVED** [review]
This looks great - I gave it a try. I was unable to bind to a Class though, i.e. I tried
```python
from napari.layers import Points
@Points.bind_key('w')
def func(layer):
    print('hello')
```
in `custom_keybindings.py` and then added a new points layer from the gui but the `w` key didn't seem to be bound. Maybe you could expand the example a little to include this use case. Maybe also modify one of the functions to print something on release too in that example. Otherwise I have no comments on the implementation, though I'll let @jni review before merging.

**Comment by royerloic** [comment]
Beautiful feature, really like it.

File: napari/components/_viewer/view/main.py
Comment (jni) [inline napari/components/_viewer/view/main.py]: What does this mean? Meaning, reworking all existing keybindings to use the new system?

Incidentally, how does this system interact with the Qt menubar system etc?

File: napari/components/_viewer/view/main.py
Comment (jni) [inline napari/components/_viewer/view/main.py]: Could you expand the names of things? `comb` -> `key_combo`, `parent` -> `key_receiver` (or somesuch, something more descriptive of what it is), `func` -> `function`, `gen` -> `result`?

File: napari/layers/_base_layer/model.py
Comment (jni) [inline napari/layers/_base_layer/model.py]: Total side note: I'm also unclear from the name `VisualWrapper` what it is or what it's doing, so that might be another thing to rename in the near future.

File: napari/layers/_image_layer/model.py
Comment (jni) [inline napari/layers/_image_layer/model.py]: Can this variable not be provided by the Mixin directly? It's a bit noisy that it has to be added to any subclass in the hierarchy...

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:31]: I love this example! :joy:

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py]: To reiterate:

```python
*modifiers, key = parsed
```

:grinning:

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:248]: What's happening here? Can you explain it in a comment?

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:276]: Maybe add a usage example to the docstring? I'm not sure how this works, how to use it, or what problem this solves, based on just reading the description.

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:304]: As above: some usage examples would be very handy here.

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py]: typo: inherited

File: napari/util/keybindings.py
Comment (jni) [inline napari/util/keybindings.py:350]: Is there really no way around this?

**Review by jni — COMMENTED** [review]
@kne42 the functionality is super awesome, and so much simpler than the last attempt! Thank you! =) I've left a few comments, but most are optional/requests for clarification. The main issue now is rebasing on master, which should be fun! =P

File: napari/layers/_base_layer/model.py
Comment (sofroniewn) [reply napari/layers/_base_layer/model.py]: `VisualWrapper` relates to our handling of the `vispy` scene and will get pulled out when we separate our vispy code from our layer models

File: napari/components/_viewer/view/main.py
Comment (kne42) [reply napari/components/_viewer/view/main.py]: Precisely.

Getting this to work with Qt's built-in shortcuts will be tricky if we want to do it natively. Otherwise, it would be pretty simple to just do something like `Viewer.bind_key('Control-W', lambda v: v.close())`

File: napari/components/_viewer/view/main.py
Comment (kne42) [reply napari/components/_viewer/view/main.py]: sure, but do you really think `function` is that much more descriptive than `func`? I thought that's pretty standard

I also chose `gen` because sometimes it returns a generator. still want to go with `result` or can you think of something that better captures this?

File: napari/layers/_image_layer/model.py
Comment (kne42) [reply napari/layers/_image_layer/model.py]: Dictionaries are mutable so it would create quite the mess if we did that. We could resolve this by using a custom metaclass but I'm not so keen on that either.

File: napari/util/keybindings.py
Comment (kne42) [reply napari/util/keybindings.py]: right, sorry I forgot this last time

File: napari/util/keybindings.py
Comment (kne42) [reply napari/util/keybindings.py:248]: it's similar to using `functools.partial` except it returns a function, which allows this to be used as a decorator with arguments and also as a regular function

File: napari/util/keybindings.py
Comment (kne42) [reply napari/util/keybindings.py:350]: Reiterating for those not following along the full conversation: we could use a metaclass which would be pretty gross too.

**Comment by kne42** [comment]
@sofroniewn this is because I forgot to update `__contains__`. I think I'd like to eventually re-implement a custom dictionary class similar to what I did with `List` because there are like 1000 more bugs with the current implementation (e.g. repr, etc.).

File: napari/components/_viewer/view/main.py
Comment (jni) [reply napari/components/_viewer/view/main.py]: @kne42 func is the least worrisome. `gen` being *sometimes* a generator is a huge problem! LOL `value_or_generator` is the only thing I can think of otherwise, but `result` works in any case. maybe add a comment, "can be a single value or a generator"?

File: napari/layers/_image_layer/model.py
Comment (jni) [reply napari/layers/_image_layer/model.py]: I was thinking something like overriding `__new__` so that if the class has a dictionary, use that, but if it doesn't, add one? Is that possible? I'm not too clear on class variables in Python actually!

File: napari/util/keybindings.py
Comment (jni) [reply napari/util/keybindings.py:248]: I'm confused. functools.partial also returns a function?

File: napari/layers/_image_layer/model.py
Comment (kne42) [reply napari/layers/_image_layer/model.py]: That's a clever way to do it but it unfortunately won't work as expected when applied to subclasses; e.g. `Pyramid` would share the same bindings as `Image`. I think we could additionally do something hacky and iterate through a class's bases, checking if one of their `class_keybindings` is the same as its own via the `is` operator. Would you like me to try that?

File: napari/util/keybindings.py
Comment (kne42) [reply napari/util/keybindings.py:248]: Ah yes, I should've used better language ;P

So I'm sure you're familiar with decorators, but to explain this aspect, it's time to go on a deep-dive *queue theme music* 😎 (let's ignore that I'm just generally bad at concisely explaining things)

So I'm sure you're familiar with simple decorators which take in a single argument, the function, do something with it, and either return the same function or a modified version. The type we're using is registration so I won't go into modifying them but you can find examples at projects like [`wrapt`](https://wrapt.readthedocs.io).

For example, here is a basic registration decorator
```python
special_func_registry = []

def special_func(func):
    special_func_registry.append(func)
    return func

@special_func
def foo(): ...
```

However, say you wanted to allow users to specify a name for the function:
```python
special_func_registry = {}

def register_special_func(name):
    def inner(func):  # this is the actual decorator
        special_func_registry[name] = func
        return func
    return inner

@register_special_func('bar')
def foo(): ...
```

What happens here is that `register_special_func('bar')` resolves to `inner` which is then applied as a decorator.

Now, say that you wanted to also use this as a regular function instead of as a decorator. You'd have to do something like `register_special_func('bar')(foo)`, which is pretty nasty.

So we revise it again:
```python
def register_special_func(name, func=None):
    if func is not None:  # used as a function
        special_func_registry[name] = func
    else:                 # used as a decorator
        def inner(func):
            register_special_func(name, func)
            return func
        return inner
```

In cases where the functional version of the decorator does not return anything, you can use `functools.partial` and have it return the decorated function:
```python
def register_special_func(name, func=None):
    if func is None:
        return functools.partial(register_special_func, name)
    special_func_registry[name] = func
    return func
```

However, if you were to try this while returning anything other than the decorated function, you would be met with disaster:
```python
def register_special_func(name, func=None):
    if func is None:
        return functools.partial(register_special_func, name)
    special_func_registry[name] = func
    return name

@register_special_func('bar')
def foo(): ...



>>> foo()
TypeError: 'str' object is not callable

>>> foo
'bar'
```

This is why I implemented it the way I did: it allows you to both use `bind_key` as a function and a decorator. As a function, `bind_key` will return any function unbound by the operation performed (whether it was by explicitly unbinding the function or overwriting it) while also not terrifyingly deleting all your functions by accident.

Comment (jni) [reply napari/util/keybindings.py:248]: ooooooh!

I *think* what you want is `toolz.curry`. SOOOO much simpler:

```python
import toolz as tz

@tz.curry
def register_special_func(name, func):
    special_func_registry[name] = func
    return func
```

Comment (kne42) [reply napari/util/keybindings.py:248]: `toolz.curry` has the same problem as using `functools.partial` as explained in the last section above, but it looks really handy, will have to try it out in future decorators!

**Comment by sofroniewn** [comment]
nice job getting this finished @kne42! our critical tests are passing, i will merge

**Pull request merged**
